cip: fix fatal bug in SplitByKnownRegistries

@thockin TL;DR: Allowing images to be pushed to the root directory currently breaks the promoter. Fix on the way...!

Pasted error log from ci-k8sio-cip run from earlier today:

```
I0304 21:47:40.563384      13 inventory.go:1615] Request {gcr.io/k8s-staging-cluster-api/kubeadm-bootstrap-controller   true}: OK
I0304 21:47:40.566394      13 inventory.go:1615] Request {us.gcr.io/k8s-artifacts-prod/resource_consumer/controller k8s-infra-gcr-promoter@k8s-artifacts-prod.iam.gserviceaccount.com  false}: OK
I0304 21:47:40.605095      13 inventory.go:1615] Request {gcr.io/google-containers/skydns-ppc64le   true}: OK
I0304 21:47:40.629320      13 inventory.go:1615] Request {us.gcr.io/k8s-artifacts-prod/federation-controller-manager-amd64 k8s-infra-gcr-promoter@k8s-artifacts-prod.iam.gserviceaccount.com  false}: OK
panic: runtime error: index out of range [0] with length 0
goroutine 162 [running]:
sigs.k8s.io/k8s-container-image-promoter/lib/dockerregistry.SplitByKnownRegistries(0xc006ecf590, 0x2f, 0xc000d74000, 0x6c, 0x92, 0x0, 0x0, 0x1, 0xd, 0x0, ...)
	lib/dockerregistry/inventory.go:1491 +0x326
sigs.k8s.io/k8s-container-image-promoter/lib/dockerregistry.(*SyncContext).ReadRegistries.func2(0xc0002ca000, 0xc0000c8240, 0xc00521b4a0, 0xc00113c800, 0xc00113c730)
	lib/dockerregistry/inventory.go:1282 +0xafa
created by sigs.k8s.io/k8s-container-image-promoter/lib/dockerregistry.(*SyncContext).ExecRequests
	lib/dockerregistry/inventory.go:1620 +0x137
```

(see https://prow.k8s.io/view/gcs/kubernetes-jenkins/logs/ci-k8sio-cip/1235320791898263552)

This happens inside the `SplitByKnownRegistries(unparsed_name, prefixes)` function, iff the registry name and the full image name path (to be split into the registry + image name) are the same.

I've reproduced this locally, and understand why this is happening. Previously (before the [backfill manifests were merged](https://github.com/kubernetes/k8s.io/pull/623), this has never happened because we've never had to read images that have an `unparsed_name` that matches a `prefix`.

Let's take an example. Currently the [us.gcr.io/k8s-artifacts-prod/kube-state-metrics](https://console.cloud.google.com/gcr/images/k8s-artifacts-prod/US/kube-state-metrics?gcrImageListsize=30) repo has some images in it as a result of the backfill. However, there is also a [promoter manifest](https://github.com/kubernetes/k8s.io/blob/679faf3610f1cbec12873795ad36f675b4007727/k8s.gcr.io/manifests/k8s-staging-kube-state-metrics/promoter-manifest.yaml#L5) that names `us.gcr.io/k8s-artifacts-prod/kube-state-metrics` as a destination repository. Normally, new images under this repository would end up as `us.gcr.io/k8s-artifacts-prod/kube-state-metrics/<NEW_IMAGE>`. And so *normally*, this would end up calling `SplitByKnownRegistries` with `us.gcr.io/k8s-artifacts-prod/kube-state-metrics/<NEW_IMAGE>` as the `unparsed_name` and `us.gcr.io/k8s-artifacts-prod/kube-state-metrics` as one of the prefixes to parse with.

However, ever since the partial backfilling that happened yesterday, there are now images that rest in this same path --- as of the time of this writing, for example, this image:

```
us.gcr.io/k8s-artifacts-prod/kube-state-metrics:v1.4.0
```

And so when we read the path `us.gcr.io/k8s-artifacts-prod/kube-state-metrics` today, we end up calling `SplitByKnownRegistries` with this name that matches one of the prefixes with the same name, `us.gcr.io/k8s-artifacts-prod/kube-state-metrics`. When https://github.com/kubernetes/k8s.io/pull/623 was *initially merged*, we did not see the same error because *no root-level images had been populated yet*. We saw the error today in https://prow.k8s.io/view/gcs/kubernetes-jenkins/logs/ci-k8sio-cip/1235320791898263552 for the first time because some images like `us.gcr.io/k8s-artifacts-prod/kube-state-metrics:v1.4.0` were created due to the partial backfill.

The fix should be simple --- if the image name matches a known prefix, just take everything after the last '/' character.

Related context:
- https://github.com/kubernetes/k8s.io/pull/623#issuecomment-594908687
- https://github.com/kubernetes/k8s.io/pull/629

--- a/cip.go
+++ b/cip.go
@@ -152,8 +152,7 @@ func main() {
 		uuid := os.Getenv("CIP_AUDIT_TESTCASE_UUID")
 		if len(uuid) > 0 {
 			klog.Infof("Starting auditor in Test Mode (%s)", uuid)
-		}
-		if uuid == "" {
+		} else {
 			uuid = guuid.New().String()
 			klog.Infof("Starting auditor in Regular Mode (%s)", uuid)
 		}

--- a/lib/dockerregistry/inventory.go
+++ b/lib/dockerregistry/inventory.go
@@ -1489,11 +1489,20 @@ func SplitByKnownRegistries(
 		if strings.HasPrefix(string(r), string(rc.Name)) {
 			trimmed := strings.TrimPrefix(string(r), string(rc.Name))
 
-			// Remove leading "/" character, if any.
-			if trimmed[0] == '/' {
+			// nolint[gocritic]
+			if trimmed == "" {
+				// The unparsed full image path `r` and rc.Name is the same ---
+				// this happens for images pushed to the root directory. Just
+				// get everything past the last '/' seen in `r` to get the image
+				// name.
+				i := strings.LastIndex(string(r), "/")
+				return rc.Name[:i], ImageName(string(r)[i+1:]), nil
+			} else if trimmed[0] == '/' {
+				// Remove leading "/" character, if any.
 				return rc.Name, ImageName(trimmed[1:]), nil
+			} else {
+				return rc.Name, ImageName(trimmed), nil
 			}
-			return rc.Name, ImageName(trimmed), nil
 		}
 	}
 

--- a/lib/dockerregistry/inventory_test.go
+++ b/lib/dockerregistry/inventory_test.go
@@ -859,6 +859,54 @@ func TestSplitRegistryImagePath(t *testing.T) {
 	}
 }
 
+func TestSplitByKnownRegistries(t *testing.T) {
+	knownRegistryNames := []RegistryName{
+		// See
+		// https://github.com/kubernetes-sigs/k8s-container-image-promoter/issues/188.
+		`us.gcr.io/k8s-artifacts-prod/kube-state-metrics`,
+	}
+	knownRegistryContexts := make([]RegistryContext, 0)
+	for _, knownRegistryName := range knownRegistryNames {
+		rc := RegistryContext{}
+		rc.Name = knownRegistryName
+		knownRegistryContexts = append(knownRegistryContexts, rc)
+	}
+
+	var tests = []struct {
+		name                 string
+		input                RegistryName
+		expectedRegistryName RegistryName
+		expectedImageName    ImageName
+		expectedErr          error
+	}{
+		{
+			`image at toplevel root path`,
+			`us.gcr.io/k8s-artifacts-prod/kube-state-metrics`,
+			`us.gcr.io/k8s-artifacts-prod`,
+			`kube-state-metrics`,
+			nil,
+		},
+	}
+	for _, test := range tests {
+		rootReg, imageName, err := SplitByKnownRegistries(test.input, knownRegistryContexts)
+		eqErr := checkEqual(rootReg, test.expectedRegistryName)
+		checkError(
+			t,
+			eqErr,
+			fmt.Sprintf("Test: `%v' failure (registry name mismatch)\n", test.input))
+		eqErr = checkEqual(imageName, test.expectedImageName)
+		checkError(
+			t,
+			eqErr,
+			fmt.Sprintf("Test: `%v' failure (image name mismatch)\n", test.input))
+		eqErr = checkEqual(err, test.expectedErr)
+		checkError(
+			t,
+			eqErr,
+			fmt.Sprintf("Test: `%v' failure (error mismatch)\n", test.input))
+	}
+}
+
 func TestCommandGeneration(t *testing.T) {
 	destRC := RegistryContext{
 		Name:           "gcr.io/foo",

